src/librustc_mir/transform/acs_propagate.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This is Alias-Constant-Simplify propagation pass. This is a composition of three distinct
+//! dataflow passes: alias-propagation, constant-propagation and terminator simplification.
+//!
+//! All these are very similar in their nature:
+//!
+//!                 | Constant  |  Alias   | Simplify  |
+//!|----------------|-----------|----------|-----------|
+//!| Lattice Domain | Lvalue    | Lvalue   | Lvalue    |
+//!| Lattice Value  | Constant  | Lvalue   | Constant  |
+//!| Transfer       | x = const | x = lval | x = const |
+//!| Rewrite        | x → const | x → lval | T(x) → T' |
+//!| Bottom         | {}        | {}       | {}        |
+//!
+//! For all of them we will be using a lattice of Hashmap from Lvalue to
+//! WTop<Either<Lvalue, Constant>>
+//!
+//! My personal believ is that it should be possible to make a way to compose two hashmap lattices
+//! into one, but I can’t seem to get it just right yet, so we do the composing and decomposing
```

Comment:
I do not think this a problem for this pass as it is currently implemented. Most notably this pass does not propagate through `Rvalue::Ref`, only through `Rvalue::Use` nor does it replace the destination lvalues. These both considered should not cause any problems related to aliased memory that do not already exist in the original MIR AFAICT.

Is there any particular examples you have in mind such pass should watch out for?